File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if q.val < p.val:
            p,q = q,p

        lca = root

        while lca:
            if q.val > lca.val and p.val > lca.val:
                lca = lca.right
            elif q.val < lca.val and p.val < lca.val:
                lca = lca.left
            else:
                return lca

        return lca

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 //bfs we like keep a stack right and then keep going
 //visit root, add left and right child to stack
 // do something to root, hten pop from stack and continue
 // until visited everything

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        

        List<List<Integer>> theList = new ArrayList<>();
        if(root == null){
            return theList;
        }


        Queue<TreeNode> q = new LinkedList<>();

        q.add(root);

        while(!q.isEmpty()){
            List<Integer> miniList = new ArrayList<>();

            for(int i = q.size(); i >0;i--){
                if(q.peek().left != null){
                    q.add(q.peek().left);
                }
                if(q.peek().right != null){
                    q.add(q.peek().right);
                }
                miniList.add(q.remove().val);
            }
            if(miniList.size()>0){
                theList.add(miniList);
            }
        }

        return theList;



    }
}
